diff --git a/build-scripts/warnings.cmake b/build-scripts/warnings.cmake
index fbe6966..80a3e02 100644
--- a/build-scripts/warnings.cmake
+++ b/build-scripts/warnings.cmake
@@ -4,7 +4,7 @@
 # global additional warnings.
 if (MSVC)
   # warning level 4
-  add_compile_options(/W4)
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/W4>)
 else ()
   # refer to https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
   add_compile_options(
diff --git a/core/iwasm/common/wasm_c_api.c b/core/iwasm/common/wasm_c_api.c
index 53d3a87..344f482 100644
--- a/core/iwasm/common/wasm_c_api.c
+++ b/core/iwasm/common/wasm_c_api.c
@@ -1273,7 +1273,7 @@ wasm_externtype_kind(const wasm_externtype_t *extern_type)
         return WASM_EXTERN_FUNC;
     }
 
-    return extern_type->extern_kind;
+    return (wasm_externkind_t)extern_type->extern_kind;
 }
 
 #define BASIC_FOUR_TYPE_LIST(V) \
@@ -3120,7 +3120,7 @@ wasm_func_new_with_env(wasm_store_t *store, const wasm_functype_t *type,
 }
 
 wasm_func_t *
-wasm_func_new_internal(wasm_store_t *store, uint16 func_idx_rt,
+wasm_func_new_internal(wasm_store_t *store, uint32 func_idx_rt,
                        WASMModuleInstanceCommon *inst_comm_rt)
 {
     wasm_func_t *func = NULL;
@@ -3190,7 +3190,7 @@ wasm_func_new_internal(wasm_store_t *store, uint16 func_idx_rt,
     func->store = store;
     func->module_name = NULL;
     func->name = NULL;
-    func->func_idx_rt = func_idx_rt;
+    func->func_idx_rt = (uint16)func_idx_rt;
     func->inst_comm_rt = inst_comm_rt;
     return func;
 
@@ -3793,7 +3793,7 @@ wasm_global_get(const wasm_global_t *global, wasm_val_t *out)
 }
 
 wasm_global_t *
-wasm_global_new_internal(wasm_store_t *store, uint16 global_idx_rt,
+wasm_global_new_internal(wasm_store_t *store, uint32 global_idx_rt,
                          WASMModuleInstanceCommon *inst_comm_rt)
 {
     wasm_global_t *global = NULL;
@@ -3868,20 +3868,20 @@ wasm_global_new_internal(wasm_store_t *store, uint16 global_idx_rt,
 
 #if WASM_ENABLE_INTERP != 0
     if (inst_comm_rt->module_type == Wasm_Module_Bytecode) {
-        interp_global_get((WASMModuleInstance *)inst_comm_rt, global_idx_rt,
-                          global->init);
+        interp_global_get((WASMModuleInstance *)inst_comm_rt,
+                          (uint16)global_idx_rt, global->init);
     }
 #endif
 
 #if WASM_ENABLE_AOT != 0
     if (inst_comm_rt->module_type == Wasm_Module_AoT) {
-        aot_global_get((AOTModuleInstance *)inst_comm_rt, global_idx_rt,
-                       global->init);
+        aot_global_get((AOTModuleInstance *)inst_comm_rt,
+                       (uint16)global_idx_rt, global->init);
     }
 #endif
 
     global->inst_comm_rt = inst_comm_rt;
-    global->global_idx_rt = global_idx_rt;
+    global->global_idx_rt = (uint16)global_idx_rt;
 
     return global;
 
@@ -3920,7 +3920,7 @@ wasm_table_new_basic(wasm_store_t *store, const wasm_tabletype_t *type)
 }
 
 wasm_table_t *
-wasm_table_new_internal(wasm_store_t *store, uint16 table_idx_rt,
+wasm_table_new_internal(wasm_store_t *store, uint32 table_idx_rt,
                         WASMModuleInstanceCommon *inst_comm_rt)
 {
     wasm_table_t *table = NULL;
@@ -3965,7 +3965,7 @@ wasm_table_new_internal(wasm_store_t *store, uint16 table_idx_rt,
     }
 
     table->inst_comm_rt = inst_comm_rt;
-    table->table_idx_rt = table_idx_rt;
+    table->table_idx_rt = (uint16)table_idx_rt;
 
     RETURN_OBJ(table, wasm_table_delete);
 }
@@ -4286,7 +4286,7 @@ wasm_memory_copy(const wasm_memory_t *src)
 }
 
 wasm_memory_t *
-wasm_memory_new_internal(wasm_store_t *store, uint16 memory_idx_rt,
+wasm_memory_new_internal(wasm_store_t *store, uint32 memory_idx_rt,
                          WASMModuleInstanceCommon *inst_comm_rt)
 {
     wasm_memory_t *memory = NULL;
@@ -4346,7 +4346,7 @@ wasm_memory_new_internal(wasm_store_t *store, uint16 memory_idx_rt,
     }
 
     memory->inst_comm_rt = inst_comm_rt;
-    memory->memory_idx_rt = memory_idx_rt;
+    memory->memory_idx_rt = (uint16)memory_idx_rt;
 
     RETURN_OBJ(memory, wasm_memory_delete);
 }
@@ -4883,7 +4883,7 @@ do_link(const wasm_instance_t *inst, const wasm_module_t *module,
 #if WASM_ENABLE_INTERP != 0
                 if ((*module)->module_type == Wasm_Module_Bytecode) {
                     ret = interp_link_func(inst, MODULE_INTERP(module),
-                                           import_func_i,
+                                           (uint16)import_func_i,
                                            wasm_extern_as_func(import));
                 }
 #endif
@@ -4907,7 +4907,7 @@ do_link(const wasm_instance_t *inst, const wasm_module_t *module,
 #if WASM_ENABLE_INTERP != 0
                 if ((*module)->module_type == Wasm_Module_Bytecode) {
                     ret = interp_link_global(MODULE_INTERP(module),
-                                             import_global_i,
+                                             (uint16)import_global_i,
                                              wasm_extern_as_global(import));
                 }
 #endif
diff --git a/core/iwasm/common/wasm_c_api_internal.h b/core/iwasm/common/wasm_c_api_internal.h
index 49a17a9..7c0fb16 100644
--- a/core/iwasm/common/wasm_c_api_internal.h
+++ b/core/iwasm/common/wasm_c_api_internal.h
@@ -226,19 +226,19 @@ wasm_foreign_new_internal(wasm_store_t *store, uint32 foreign_idx_rt,
                           WASMModuleInstanceCommon *inst_comm_rt);
 
 wasm_func_t *
-wasm_func_new_internal(wasm_store_t *store, uint16 func_idx_rt,
+wasm_func_new_internal(wasm_store_t *store, uint32 func_idx_rt,
                        WASMModuleInstanceCommon *inst_comm_rt);
 
 wasm_global_t *
-wasm_global_new_internal(wasm_store_t *store, uint16 global_idx_rt,
+wasm_global_new_internal(wasm_store_t *store, uint32 global_idx_rt,
                          WASMModuleInstanceCommon *inst_comm_rt);
 
 wasm_memory_t *
-wasm_memory_new_internal(wasm_store_t *store, uint16 memory_idx_rt,
+wasm_memory_new_internal(wasm_store_t *store, uint32 memory_idx_rt,
                          WASMModuleInstanceCommon *inst_comm_rt);
 
 wasm_table_t *
-wasm_table_new_internal(wasm_store_t *store, uint16 table_idx_rt,
+wasm_table_new_internal(wasm_store_t *store, uint32 table_idx_rt,
                         WASMModuleInstanceCommon *inst_comm_rt);
 
 void
diff --git a/core/iwasm/common/wasm_exec_env.c b/core/iwasm/common/wasm_exec_env.c
index 4775295..0ea4f23 100644
--- a/core/iwasm/common/wasm_exec_env.c
+++ b/core/iwasm/common/wasm_exec_env.c
@@ -91,6 +91,8 @@ wasm_exec_env_create_internal(struct WASMModuleInstanceCommon *module_inst,
 
     return exec_env;
 
+#if defined(OS_ENABLE_HW_BOUND_CHECK) || WASM_ENABLE_THREAD_MGR != 0 \
+    || WASM_ENABLE_AOT != 0
 #ifdef OS_ENABLE_HW_BOUND_CHECK
 fail5:
 #if WASM_ENABLE_THREAD_MGR != 0 && WASM_ENABLE_DEBUG_INTERP != 0
@@ -112,6 +114,7 @@ fail1:
 #endif
     wasm_runtime_free(exec_env);
     return NULL;
+#endif
 }
 
 void
diff --git a/core/iwasm/common/wasm_runtime_common.c b/core/iwasm/common/wasm_runtime_common.c
index 55aadc3..f9b7637 100644
--- a/core/iwasm/common/wasm_runtime_common.c
+++ b/core/iwasm/common/wasm_runtime_common.c
@@ -6854,7 +6854,7 @@ interp_mark_all_externrefs(WASMModuleInstance *module_inst)
         if (elem_type == VALUE_TYPE_EXTERNREF) {
             table_data = table->elems;
             for (j = 0; j < table->cur_size; j++) {
-                externref_idx = table_data[j];
+                externref_idx = (uint32)table_data[j];
                 mark_externref(externref_idx);
             }
         }
diff --git a/core/iwasm/common/wasm_shared_memory.c b/core/iwasm/common/wasm_shared_memory.c
index c1aa18f..e8410fe 100644
--- a/core/iwasm/common/wasm_shared_memory.c
+++ b/core/iwasm/common/wasm_shared_memory.c
@@ -293,13 +293,13 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
               || module->module_type == Wasm_Module_AoT);
 
     if (wasm_copy_exception(module_inst, NULL)) {
-        return -1;
+        return (uint32)-1;
     }
 
     /* Currently we have only one memory instance */
     if (!shared_memory_is_shared(module_inst->memories[0])) {
         wasm_runtime_set_exception(module, "expected shared memory");
-        return -1;
+        return (uint32)-1;
     }
 
     shared_memory_lock(module_inst->memories[0]);
@@ -316,7 +316,7 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
                 > module_inst->memories[0]->memory_data_end)) {
         shared_memory_unlock(module_inst->memories[0]);
         wasm_runtime_set_exception(module, "out of bounds memory access");
-        return -1;
+        return (uint32)-1;
     }
     shared_memory_unlock(module_inst->memories[0]);
 
@@ -343,7 +343,7 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
     if (!(wait_node = wasm_runtime_malloc(sizeof(AtomicWaitNode)))) {
         os_mutex_unlock(lock);
         wasm_runtime_set_exception(module, "failed to create wait node");
-        return -1;
+        return (uint32)-1;
     }
     memset(wait_node, 0, sizeof(AtomicWaitNode));
 
@@ -351,7 +351,7 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
         os_mutex_unlock(lock);
         wasm_runtime_free(wait_node);
         wasm_runtime_set_exception(module, "failed to init wait cond");
-        return -1;
+        return (uint32)-1;
     }
 
     wait_node->status = S_WAITING;
@@ -364,7 +364,7 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
         os_cond_destroy(&wait_node->wait_cond);
         wasm_runtime_free(wait_node);
         wasm_runtime_set_exception(module, "failed to acquire wait_info");
-        return -1;
+        return (uint32)-1;
     }
 
     /* unit of timeout is nsec, convert it to usec */
@@ -448,7 +448,7 @@ wasm_runtime_atomic_notify(WASMModuleInstanceCommon *module, void *address,
 
     if (out_of_bounds) {
         wasm_runtime_set_exception(module, "out of bounds memory access");
-        return -1;
+        return (uint32)-1;
     }
 
     /* Currently we have only one memory instance */
diff --git a/core/iwasm/interpreter/wasm.h b/core/iwasm/interpreter/wasm.h
index c60349d..161fc65 100644
--- a/core/iwasm/interpreter/wasm.h
+++ b/core/iwasm/interpreter/wasm.h
@@ -1324,7 +1324,8 @@ wasm_value_type_size_internal(uint8 value_type, uint8 pointer_size)
 inline static uint16
 wasm_value_type_cell_num_internal(uint8 value_type, uint8 pointer_size)
 {
-    return wasm_value_type_size_internal(value_type, pointer_size) / 4;
+    return (uint16)(wasm_value_type_size_internal(value_type, pointer_size)
+                    / 4);
 }
 
 /**
@@ -1339,7 +1340,7 @@ wasm_value_type_size(uint8 value_type)
 inline static uint16
 wasm_value_type_cell_num(uint8 value_type)
 {
-    return wasm_value_type_size(value_type) / 4;
+    return (uint16)(wasm_value_type_size(value_type) / 4);
 }
 
 inline static uint32
@@ -1508,7 +1509,6 @@ block_type_get_arity(const BlockType *block_type, uint8 label_type)
         else
             return block_type->u.type->result_count;
     }
-    return 0;
 }
 
 #ifdef __cplusplus
diff --git a/core/iwasm/interpreter/wasm_interp_fast.c b/core/iwasm/interpreter/wasm_interp_fast.c
index 91ed17d..6d3578f 100644
--- a/core/iwasm/interpreter/wasm_interp_fast.c
+++ b/core/iwasm/interpreter/wasm_interp_fast.c
@@ -428,7 +428,7 @@ wasm_interp_get_frame_ref(WASMInterpFrame *frame)
             local_type = cur_func->local_types[local_idx - param_count]; \
     } while (0)
 
-#define GET_OFFSET() (frame_ip += 2, LOAD_I16(frame_ip - 2))
+#define GET_OFFSET() LOAD_I16((frame_ip += 2) - 2)
 
 #define SET_OPERAND_I32(off, value)                                 \
     do {                                                            \
@@ -1977,7 +1977,7 @@ wasm_interp_call_func_bytecode(WASMModuleInstance *module,
                 }
 
 #if WASM_ENABLE_GC == 0
-                PUSH_I32(tbl_inst->elems[elem_idx]);
+                PUSH_I32((int32)tbl_inst->elems[elem_idx]);
 #else
                 PUSH_REF(tbl_inst->elems[elem_idx]);
 #endif
@@ -4006,7 +4006,7 @@ wasm_interp_call_func_bytecode(WASMModuleInstance *module,
                 /* TODO: multi-memory wasm_enlarge_memory_with_idx() */
                 if (!wasm_enlarge_memory(module, delta)) {
                     /* failed to memory.grow, return -1 */
-                    frame_lp[addr_ret] = -1;
+                    frame_lp[addr_ret] = (uint32)-1;
                 }
                 else {
                     /* success, return previous page count */
@@ -5283,7 +5283,7 @@ wasm_interp_call_func_bytecode(WASMModuleInstance *module,
                         uint8 fill_val, *mdst;
 
                         len = POP_I32();
-                        fill_val = POP_I32();
+                        fill_val = (uint8)POP_I32();
                         dst = POP_I32();
 
 #if WASM_ENABLE_THREAD_MGR != 0
diff --git a/core/iwasm/interpreter/wasm_loader.c b/core/iwasm/interpreter/wasm_loader.c
index 8ca8333..613d959 100644
--- a/core/iwasm/interpreter/wasm_loader.c
+++ b/core/iwasm/interpreter/wasm_loader.c
@@ -2908,7 +2908,8 @@ load_table_import(const uint8 **p_buf, const uint8 *buf_end,
         return false;
     }
 
-    if (!wasm_table_check_flags(table_flag, error_buf, error_buf_size, false)) {
+    if (!wasm_table_check_flags((uint8)table_flag, error_buf, error_buf_size,
+                                false)) {
         return false;
     }
 
@@ -2976,9 +2977,9 @@ load_table_import(const uint8 **p_buf, const uint8 *buf_end,
     }
 
     /* now we believe all declaration are ok */
-    table->table_type.elem_type = declare_elem_type;
+    table->table_type.elem_type = (uint8)declare_elem_type;
     table->table_type.init_size = declare_init_size;
-    table->table_type.flags = table_flag;
+    table->table_type.flags = (uint8)table_flag;
     table->table_type.max_size = declare_max_size;
 
 #if WASM_ENABLE_WAMR_COMPILER != 0
@@ -3076,7 +3077,8 @@ load_memory_import(const uint8 **p_buf, const uint8 *buf_end,
         return false;
     }
 
-    if (!wasm_memory_check_flags(mem_flag, error_buf, error_buf_size, false)) {
+    if (!wasm_memory_check_flags((uint8)mem_flag, error_buf, error_buf_size,
+                                 false)) {
         return false;
     }
 
@@ -3387,6 +3389,7 @@ load_table(const uint8 **p_buf, const uint8 *buf_end, WASMModule *module,
     bool need_ref_type_map;
 #endif
     bool is_table64 = false;
+    uint32 table_flag;
 
 #if WASM_ENABLE_GC == 0
     CHECK_BUF(p, p_end, 1);
@@ -3432,7 +3435,8 @@ load_table(const uint8 **p_buf, const uint8 *buf_end, WASMModule *module,
 #endif /* end of WASM_ENABLE_GC == 0 */
 
     p_org = p;
-    read_leb_uint32(p, p_end, table->table_type.flags);
+    read_leb_uint32(p, p_end, table_flag);
+    table->table_type.flags = (uint8)table_flag;
     is_table64 = table->table_type.flags & TABLE64_FLAG;
     if (p - p_org > 1) {
         LOG_VERBOSE("integer representation too long(table)");
@@ -3492,8 +3496,8 @@ load_memory(const uint8 **p_buf, const uint8 *buf_end, WASMMemory *memory,
         return false;
     }
 
-    if (!wasm_memory_check_flags(memory->flags, error_buf, error_buf_size,
-                                 false)) {
+    if (!wasm_memory_check_flags((uint8)memory->flags, error_buf,
+                                 error_buf_size, false)) {
         return false;
     }
 
@@ -3849,7 +3853,8 @@ init_function_local_offsets(WASMFunction *func, char *error_buf,
         local_offset += wasm_value_type_cell_num(local_types[i]);
     }
 
-    bh_assert(local_offset == func->param_cell_num + func->local_cell_num);
+    bh_assert(local_offset
+              == (uint32)func->param_cell_num + func->local_cell_num);
     return true;
 }
 
@@ -4805,7 +4810,7 @@ load_init_expr_vec(const uint8 **p_buf, const uint8 *buf_end,
         InitializerExpression *init_expr = &table_segment->init_values[i];
 
         if (!load_init_expr(module, &p, p_end, init_expr,
-                            table_segment->elem_type,
+                            (uint8)table_segment->elem_type,
 #if WASM_ENABLE_GC == 0
                             NULL,
 #else
@@ -6901,6 +6906,9 @@ create_module(char *name, char *error_buf, uint32 error_buf_size)
     (void)ret;
     return module;
 
+#if WASM_ENABLE_GC != 0 || WASM_ENABLE_DEBUG_INTERP != 0        \
+    || (WASM_ENABLE_FAST_JIT != 0 && WASM_ENABLE_JIT != 0 \
+        && WASM_ENABLE_LAZY_JIT != 0)
 #if WASM_ENABLE_DEBUG_INTERP != 0                    \
     || (WASM_ENABLE_FAST_JIT != 0 && WASM_ENABLE_JIT \
         && WASM_ENABLE_LAZY_JIT != 0)
@@ -6914,6 +6922,7 @@ fail1:
 #endif
     wasm_runtime_free(module);
     return NULL;
+#endif
 }
 
 #if WASM_ENABLE_DEBUG_INTERP != 0
@@ -9731,7 +9740,7 @@ preserve_referenced_local(WASMLoaderContext *loader_ctx, uint8 opcode,
                 if (loader_ctx->p_code_compiled) {
                     bh_assert(preserved_offset != (int16)local_index);
                 }
-                if (is_32bit_type(local_type)) {
+                if (is_32bit_type((uint8)local_type)) {
                     /* Only increase preserve offset in the second traversal */
                     if (loader_ctx->p_code_compiled)
                         loader_ctx->preserved_local_offset++;
@@ -9758,7 +9767,7 @@ preserve_referenced_local(WASMLoaderContext *loader_ctx, uint8 opcode,
                     return false;
                 }
 
-                emit_operand(loader_ctx, local_index);
+                emit_operand(loader_ctx, (int16)local_index);
                 emit_operand(loader_ctx, preserved_offset);
                 emit_label(opcode);
             }
@@ -9942,11 +9951,11 @@ wasm_loader_emit_br_info(WASMLoaderContext *ctx, BranchBlock *frame_csp,
         if (frame_csp->label_type == LABEL_TYPE_LOOP)
             /* Use start_dynamic_offset which was set in
                copy_params_to_dynamic_space */
-            dynamic_offset = frame_csp->start_dynamic_offset
-                             + wasm_get_cell_num(types, arity);
+            dynamic_offset = (uint16)(frame_csp->start_dynamic_offset
+                                      + wasm_get_cell_num(types, arity));
         else
-            dynamic_offset =
-                frame_csp->dynamic_offset + wasm_get_cell_num(types, arity);
+            dynamic_offset = (uint16)(frame_csp->dynamic_offset
+                                      + wasm_get_cell_num(types, arity));
         if (is_br)
             ctx->dynamic_offset = dynamic_offset;
         for (i = (int32)arity - 1; i >= 0; i--) {
@@ -10072,7 +10081,7 @@ wasm_loader_pop_frame_offset(WASMLoaderContext *ctx, uint8 type,
     if (check_dynamic_offset_pop(ctx, cell_num_to_pop)
         && (*(ctx->frame_offset) > ctx->start_dynamic_offset)
         && (*(ctx->frame_offset) < ctx->max_dynamic_offset))
-        ctx->dynamic_offset -= cell_num_to_pop;
+        ctx->dynamic_offset -= (int16)cell_num_to_pop;
 
     emit_operand(ctx, *(ctx->frame_offset));
 
@@ -10267,8 +10276,9 @@ wasm_loader_get_const_offset(WASMLoaderContext *ctx, uint8 type, void *value,
             }
 
             /* constant index is encoded as negative value */
-            *offset = -(int32)(ctx->i64_const_num * 2 + ctx->i32_const_num)
-                      + (int32)(i64_const - ctx->i64_consts) * 2;
+            *offset = (int16)(-(int32)(ctx->i64_const_num * 2
+                                       + ctx->i32_const_num)
+                              + (int32)(i64_const - ctx->i64_consts) * 2);
         }
         else if (type == VALUE_TYPE_V128) {
             V128 key = *(V128 *)value, *v128_const;
@@ -10280,8 +10290,8 @@ wasm_loader_get_const_offset(WASMLoaderContext *ctx, uint8 type, void *value,
             }
 
             /* constant index is encoded as negative value */
-            *offset = -(int32)(ctx->v128_const_num)
-                      + (int32)(v128_const - ctx->v128_consts);
+            *offset = (int16)(-(int32)(ctx->v128_const_num)
+                              + (int32)(v128_const - ctx->v128_consts));
         }
 
         else {
@@ -10294,8 +10304,8 @@ wasm_loader_get_const_offset(WASMLoaderContext *ctx, uint8 type, void *value,
             }
 
             /* constant index is encoded as negative value */
-            *offset = -(int32)(ctx->i32_const_num)
-                      + (int32)(i32_const - ctx->i32_consts);
+            *offset = (int16)(-(int32)(ctx->i32_const_num)
+                              + (int32)(i32_const - ctx->i32_consts));
         }
 
         return true;
@@ -10526,8 +10536,8 @@ reserve_block_ret(WASMLoaderContext *loader_ctx, uint8 opcode,
      *   Part e: each value's dst offset
      */
     frame_offset = frame_offset_org = loader_ctx->frame_offset;
-    dynamic_offset = dynamic_offset_org =
-        block->dynamic_offset + wasm_get_cell_num(return_types, return_count);
+    dynamic_offset = dynamic_offset_org = (int16)(
+        block->dynamic_offset + wasm_get_cell_num(return_types, return_count));
 
     /* First traversal to get the count of values needed to be copied. */
     for (i = (int32)return_count - 1; i >= 0; i--) {
@@ -10637,7 +10647,7 @@ reserve_block_ret(WASMLoaderContext *loader_ctx, uint8 opcode,
             if (opcode == WASM_OP_ELSE) {
                 if (cell == 4) {
                     for (int k = 0; k < cell; k++) {
-                        *(frame_offset + k) = dynamic_offset + k;
+                        *(frame_offset + k) = (int16)(dynamic_offset + k);
                     }
                 }
                 else {
@@ -11656,7 +11666,7 @@ get_table_seg_elem_type(const WASMModule *module, uint32 table_seg_idx,
     }
 
     if (p_elem_type) {
-        *p_elem_type = module->table_segments[table_seg_idx].elem_type;
+        *p_elem_type = (uint8)module->table_segments[table_seg_idx].elem_type;
     }
 #if WASM_ENABLE_GC != 0
     if (p_elem_ref_type)
@@ -13482,7 +13492,8 @@ re_scan:
 #if WASM_ENABLE_REF_TYPES != 0 || WASM_ENABLE_GC != 0
             case WASM_OP_SELECT_T:
             {
-                uint8 vec_len, type;
+                uint32 vec_len;
+                uint8 type;
 #if WASM_ENABLE_GC != 0
                 WASMRefType *ref_type = NULL;
 #endif
diff --git a/core/iwasm/interpreter/wasm_runtime.c b/core/iwasm/interpreter/wasm_runtime.c
index b1ef12d..71ca00b 100644
--- a/core/iwasm/interpreter/wasm_runtime.c
+++ b/core/iwasm/interpreter/wasm_runtime.c
@@ -849,7 +849,7 @@ functions_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
         function->local_offsets = function->u.func->local_offsets;
 
 #if WASM_ENABLE_FAST_INTERP != 0
-        function->const_cell_num = function->u.func->const_cell_num;
+        function->const_cell_num = (uint16)function->u.func->const_cell_num;
 #endif
 
         function++;
diff --git a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
index 949e574..b61ab4a 100644
--- a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
+++ b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
@@ -158,7 +158,7 @@ bh_sockaddr_to_wasi_addr(const bh_sockaddr_t *sockaddr,
         wasi_addr->addr.ip4.addr.n0 =
             (sockaddr->addr_buffer.ipv4 & 0xFF000000) >> 24;
         wasi_addr->addr.ip4.addr.n1 =
-            (sockaddr->addr_buffer.ipv4 & 0x00FF0000) >> 16;
+            (uint8_t)((sockaddr->addr_buffer.ipv4 & 0x00FF0000) >> 16);
         wasi_addr->addr.ip4.addr.n2 =
             (sockaddr->addr_buffer.ipv4 & 0x0000FF00) >> 8;
         wasi_addr->addr.ip4.addr.n3 = (sockaddr->addr_buffer.ipv4 & 0x000000FF);
@@ -2556,7 +2556,7 @@ wasi_ssp_sock_addr_resolve(wasm_exec_env_t exec_env, struct fd_table *curfds,
         return convert_errno(errno);
     }
 
-    *max_info_size = _max_info_size;
+    *max_info_size = (__wasi_size_t)_max_info_size;
     actual_info_size =
         addr_info_size < *max_info_size ? addr_info_size : *max_info_size;
 
@@ -2872,8 +2872,8 @@ wasmtime_ssp_sock_recv_from(wasm_exec_env_t exec_env, struct fd_table *curfds,
     }
 
     /* Consume bh_sockaddr_t instead of __wasi_addr_t */
-    ret = blocking_op_socket_recv_from(exec_env, fo->file_handle, buf, buf_len,
-                                       0, sockaddr_ptr);
+    ret = blocking_op_socket_recv_from(exec_env, fo->file_handle, buf,
+                                       (unsigned int)buf_len, 0, sockaddr_ptr);
     fd_object_release(exec_env, fo);
     if (-1 == ret) {
         return convert_errno(errno);
@@ -2903,7 +2903,7 @@ wasmtime_ssp_sock_send(wasm_exec_env_t exec_env, struct fd_table *curfds,
         return error;
     }
 
-    ret = os_socket_send(fo->file_handle, buf, buf_len);
+    ret = os_socket_send(fo->file_handle, buf, (unsigned int)buf_len);
     fd_object_release(exec_env, fo);
     if (-1 == ret) {
         return convert_errno(errno);
@@ -2942,8 +2942,8 @@ wasmtime_ssp_sock_send_to(wasm_exec_env_t exec_env, struct fd_table *curfds,
     wasi_addr_to_bh_sockaddr(dest_addr, &sockaddr);
 
     /* Consume bh_sockaddr instead of __wasi_addr_t */
-    ret = blocking_op_socket_send_to(exec_env, fo->file_handle, buf, buf_len, 0,
-                                     &sockaddr);
+    ret = blocking_op_socket_send_to(exec_env, fo->file_handle, buf,
+                                     (unsigned int)buf_len, 0, &sockaddr);
     fd_object_release(exec_env, fo);
     if (-1 == ret) {
         return convert_errno(errno);
@@ -3154,7 +3154,7 @@ init_address_mask(uint8_t *buf, size_t buflen, size_t mask)
         }
         else {
             size_t offset = min(mask - i * element_size, element_size);
-            buf[i] = (~0u) << (element_size - offset);
+            buf[i] = (uint8_t)((~0u) << (element_size - offset));
         }
     }
 }
diff --git a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/random.c b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/random.c
index 665a9de..acb9549 100644
--- a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/random.c
+++ b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/random.c
@@ -135,7 +135,7 @@ random_uniform(uintmax_t upper, uintmax_t *out)
     // Compute 2^k % upper
     //      == (2^k - upper) % upper
     //      == -upper % upper.
-    uintmax_t lower = -upper % upper;
+    uintmax_t lower = (0 - upper) % upper;
     for (;;) {
         uintmax_t value;
         __wasi_errno_t error = random_buf(&value, sizeof(value));
diff --git a/core/shared/mem-alloc/ems/ems_kfc.c b/core/shared/mem-alloc/ems/ems_kfc.c
index e761cb6..da243a9 100644
--- a/core/shared/mem-alloc/ems/ems_kfc.c
+++ b/core/shared/mem-alloc/ems/ems_kfc.c
@@ -217,12 +217,8 @@ adjust_ptr(uint8 **p_ptr, intptr_t offset)
      * when p_ptr is over 0x8000000000000000  by not using
      * `(intptr_t)`
      */
-    uintptr_t offset_val = 0;
-#if UINTPTR_MAX == UINT64_MAX
-    offset_val = labs(offset);
-#else
-    offset_val = abs(offset);
-#endif
+    uintptr_t offset_val = offset < 0 ? (uintptr_t)0 - (uintptr_t)offset
+                                       : (uintptr_t)offset;
 
     if (offset > 0) {
         *p_ptr = (uint8 *)((uintptr_t)(*p_ptr) + offset_val);
diff --git a/core/shared/platform/windows/win_socket.c b/core/shared/platform/windows/win_socket.c
index 8d61c45..2fb94ef 100644
--- a/core/shared/platform/windows/win_socket.c
+++ b/core/shared/platform/windows/win_socket.c
@@ -112,7 +112,7 @@ os_socket_bind(bh_socket_t socket, const char *host, int *port)
     assert(port);
 
     addr.sin_addr.s_addr = inet_addr(host);
-    addr.sin_port = htons(*port);
+    addr.sin_port = htons((u_short)*port);
     addr.sin_family = AF_INET;
 
     ret = bind(socket->raw.socket, (struct sockaddr *)&addr, sizeof(addr));
@@ -247,7 +247,10 @@ os_socket_close(bh_socket_t socket)
 __wasi_errno_t
 os_socket_shutdown(bh_socket_t socket)
 {
-    CHECK_VALID_SOCKET_HANDLE(socket);
+    if (socket == NULL)
+        return __WASI_EBADF;
+    if (socket->type != windows_handle_type_socket)
+        return __WASI_ENOTSOCK;
 
     if (shutdown(socket->raw.socket, SD_BOTH) != 0) {
         return convert_winsock_error_code(WSAGetLastError());
diff --git a/core/shared/platform/windows/win_thread.c b/core/shared/platform/windows/win_thread.c
index 1f6a57e..162e924 100644
--- a/core/shared/platform/windows/win_thread.c
+++ b/core/shared/platform/windows/win_thread.c
@@ -265,7 +265,7 @@ static unsigned __stdcall os_thread_wrapper(void *arg)
     os_mutex_unlock(&parent->wait_lock);
 
     if (!result)
-        return -1;
+        return (unsigned)-1;
 
     retval = thread_data->start_routine(thread_data->arg);
 
@@ -790,7 +790,7 @@ GetCurrentThreadStackLimits_Win7(PULONG_PTR p_low_limit,
 
     if (!tib) {
         os_printf("warning: NtCurrentTeb() failed\n");
-        return -1;
+        return (ULONG)-1;
     }
 
     *p_high_limit = (ULONG_PTR)tib->StackBase;
